torch/testing/_internal/common_methods_invocations.py

```diff
@@ -2288,6 +2288,32 @@ def generator():
 
     return list(generator())
 
+def sample_inputs_layer_norm(opinfo, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    # Ordered as input shape, normalized_shape and a kwarg dict for eps
+    cases: Tuple[Tuple[int], Tuple[int], dict] = (  # type: ignore[assignment]
+        ((1, 2, 3), (1, 2, 3), {'eps': 0.5}),
+        ((1, 2, 3), (2, 3), {'eps': -0.5}),
+        ((1,), (1,), {}),
+        ((1, 2), (2,), {}),
+    )
+
+    def generator():
+        for input_shape, normalized_shape, kwargs in cases:
+            # Shape of weight and bias should be the same as normalized_shape
+            weight = make_arg(normalized_shape)
+            bias = make_arg(normalized_shape)
+            yield SampleInput(
+                make_arg(input_shape),
+                args=(normalized_shape, weight, bias),
+                kwargs=kwargs
+            )
```

Comment:
This is a great question. Thanks! The reason I think this can't be done in this PR is:

* With `None` passed as inputs, `to_numpy` function in `SampleInput` class (used for `test_reference_testing`) doesn't accept `None` type inputs. When I created this function, I wasn't aware of this use case. I've added a TODO comment for me, have locally tested the fix, and will create a separate PR for this (to track changes in the future).

Also, after locally fixing `to_numpy` function for `None` inputs, tests passed for (weight, `None` bias) but a couple tests failed (`None` weight, bias):

<details><summary>Errors (click me)</summary>
<p>

```python
======================================================================
ERROR: test_fn_gradgrad_nn_functional_layer_norm_cpu_float64 (__main__.TestGradientsCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/Documents/pytorch/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
    raise rte
  File "/home/krshrimali/Documents/pytorch/torch/testing/_internal/common_device_type.py", line 373, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/krshrimali/Documents/pytorch/torch/testing/_internal/common_device_type.py", line 780, in test_wrapper
    return test(*args, **kwargs)
  File "/home/krshrimali/Documents/pytorch/test/test_ops.py", line 616, in test_fn_gradgrad
    self._gradgrad_test_helper(device, dtype, op, op.get_op())
  File "/home/krshrimali/Documents/pytorch/test/test_ops.py", line 582, in _gradgrad_test_helper
    return self._check_helper(device, dtype, op, variant, 'gradgradcheck')
  File "/home/krshrimali/Documents/pytorch/test/test_ops.py", line 563, in _check_helper
    self.assertTrue(gradgradcheck(fn, gradcheck_args,
  File "/home/krshrimali/Documents/pytorch/torch/testing/_internal/common_utils.py", line 2479, in gradgradcheck
    return torch.autograd.gradgradcheck(fn, inputs, grad_outputs, **kwargs)
  File "/home/krshrimali/Documents/pytorch/torch/autograd/gradcheck.py", line 1386, in gradgradcheck
    return gradcheck(
  File "/home/krshrimali/Documents/pytorch/torch/autograd/gradcheck.py", line 1262, in gradcheck
    return _gradcheck_helper(**args)
  File "/home/krshrimali/Documents/pytorch/torch/autograd/gradcheck.py", line 1275, in _gradcheck_helper
    _gradcheck_real_imag(gradcheck_fn, func, func_out, tupled_inputs, outputs, eps,
  File "/home/krshrimali/Documents/pytorch/torch/autograd/gradcheck.py", line 946, in _gradcheck_real_imag
    gradcheck_fn(func, func_out, tupled_inputs, outputs, eps,
  File "/home/krshrimali/Documents/pytorch/torch/autograd/gradcheck.py", line 1167, in _fast_gradcheck
    _check_analytical_numerical_equal(analytical_vJu, numerical_vJu, complex_indices,
  File "/home/krshrimali/Documents/pytorch/torch/autograd/gradcheck.py", line 1147, in _check_analytical_numerical_equal
    raise GradcheckError(_get_notallclose_msg(a, n, j, i, complex_indices, test_imag, is_forward_ad) + jacobians_str)
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 1 with respect to input 2,
numerical:tensor(0.9981)
analytical:tensor(0.)

The above quantities relating the numerical and analytical jacobians are computed 
in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background 
about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:

Numerical:
 tensor([[1.0000, 0.0000],
        [0.0000, 1.0000]])
Analytical:
tensor([[0., 0.],
        [0., 0.]])

The max per-element difference (slow mode) is: 1.0000000000000009.


======================================================================
ERROR: test_variant_consistency_jit_nn_functional_layer_norm_cpu_float32 (__main__.TestJitCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/krshrimali/Documents/pytorch/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
    raise rte
  File "/home/krshrimali/Documents/pytorch/torch/testing/_internal/common_device_type.py", line 373, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/krshrimali/Documents/pytorch/torch/testing/_internal/common_device_type.py", line 780, in test_wrapper
    return test(*args, **kwargs)
  File "/home/krshrimali/Documents/pytorch/test/test_ops.py", line 744, in test_variant_consistency_jit
    check_against_reference(self,
  File "/home/krshrimali/Documents/pytorch/torch/testing/_internal/common_jit.py", line 123, in check_against_reference
    grads2 = torch.autograd.grad(l2, recording_tensors, allow_unused=allow_unused)
  File "/home/krshrimali/Documents/pytorch/torch/autograd/__init__.py", line 233, in grad
    return Variable._execution_engine.run_backward(
RuntimeError: Expected a Tensor of type Variable but found an undefined Tensor for argument #1 'other'

----------------------------------------------------------------------
Ran 29 tests in 0.550s
```
</p>
</details>

Is this because `layer_norm` is not differentiable w.r.t bias?

Reply:
Er, I feel like the analytical one is right. The bias shouldn't show up in the second derivative computation so the right value is all zeros. cc @albanD @soulitzer could you check me on this?

Reply:
In any case, this failure is pre-existing and we are not enabling that test case in this PR (because we need to fix the to_numpy method to take None inputs in a follow-up), so we don't have to resolve this question now. But it would be good to try to get a minimal repro of what is going on here and file an issue around it.